src/PricingModules/UniswapV3/UniswapV3PricingModule.sol

```diff
@@ -0,0 +1,362 @@
+        (,, address token0, address token1, uint24 fee, int24 tickLower, int24 tickUpper, uint128 liquidity,,,,) =
+            INonfungiblePositionManager(asset).positions(assetId);
+
+        {
+            IUniswapV3Pool pool =
+                IUniswapV3Pool(PoolAddress.computeAddress(assetToV3Factory[asset], token0, token1, fee));
+
+            // We calculate current tick via the TWAP price. TWAP prices can be manipulated, but it is costly (not atomic).
+            // We do not use the TWAP price to calculate the current value of the asset, only to ensure ensure that the deposited Liquidity Range
+            // hence the risk of manipulation is acceptable since it can never be used to steal funds (only to deposit ranges further than 5x).
+            int24 tickCurrent = _getTwat(pool);
+
+            // The liquidity must be in an acceptable range (from 0.2x to 5X the current price).
+            // Tick difference defined as: (sqrt(1.0001))log(sqrt(5)) = 16095.2
+            require(tickCurrent - tickLower <= 16_095, "PMUV3_PD: Tlow not in limits");
+            require(tickUpper - tickCurrent <= 16_095, "PMUV3_PD: Tup not in limits");
+        }
+
+        // Cache sqrtRatio.
+        uint160 sqrtRatioLowerX96 = TickMath.getSqrtRatioAtTick(tickLower);
+        uint160 sqrtRatioUpperX96 = TickMath.getSqrtRatioAtTick(tickUpper);
+
+        // Calculate the maximal possible exposure to each underlying asset.
+        uint256 amount0Max = LiquidityAmounts.getAmount0ForLiquidity(sqrtRatioLowerX96, sqrtRatioUpperX96, liquidity);
+        uint256 amount1Max = LiquidityAmounts.getAmount1ForLiquidity(sqrtRatioLowerX96, sqrtRatioUpperX96, liquidity);
+
+        // Calculate updated exposure.
+        uint256 exposure0 = amount0Max + exposure[token0].exposure;
+        uint256 exposure1 = amount1Max + exposure[token1].exposure;
+
+        // Check that exposure doesn't exceed maxExposure
+        require(exposure0 <= exposure[token0].maxExposure, "PMUV3_PD: Exposure0 not in limits");
+        require(exposure1 <= exposure[token1].maxExposure, "PMUV3_PD: Exposure1 not in limits");
+
+        // Update exposure
+        // Unsafe casts: we already know from previous requires that exposure is smaller as maxExposure (uint128).
+        exposure[token0].exposure = uint128(exposure0);
+        exposure[token1].exposure = uint128(exposure1);
+    }
+
+    /**
+     * @notice Calculates the time weighted average tick over 300s.
+     * @param pool The liquidity pool.
+     * @return tick The time weighted average tick over 300s.
+     */
+    function _getTwat(IUniswapV3Pool pool) internal view returns (int24 tick) {
+        uint32[] memory secondsAgos = new uint32[](2);
+        secondsAgos[1] = 300; // We take a 5 minute time interval.
+
+        (int56[] memory tickCumulatives,) = pool.observe(secondsAgos);
+
+        tick = int24((tickCumulatives[0] - tickCumulatives[1]) / 300);
```

Comment:
what are the reasonable values in tickCumulatives, as it's unsafely casted?

Reply:
Can never over or underflow, check tests